net/protocol/replicator.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2025 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package protocol
+
+import (
+	"context"
+	"errors"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/sourcenetwork/defradb/event"
+	"github.com/sourcenetwork/defradb/net/message"
+)
+
+// PushLogRequest is the struct used to send a resource update to a peer node
+type PushLogRequest struct {
+	message.MetaData
+	DocID        string
+	CID          []byte
+	CollectionID string
+	Creator      string
+	Block        []byte
+}
+
+// PushLogReply is the expected response struct that should be received after
+// an pushlog request.
+type PushLogReply struct {
+	message.MetaData
+}
+
+const (
+	replicatorProtocolRequest  = "/defradb/rep_req/0.0.1"
+	replicatorProtocolResponse = "/defradb/rep_resp/0.0.1"
+)
+
+type pushLogFunc func(
+	ctx context.Context,
+	req *PushLogRequest,
+	isReplicator bool,
+) (*PushLogReply, error)
+
+type replicatorFailureFunc func(ctx context.Context, peerID, docID string) error
+
+// ReplicatorProtocol is the protocol implementation for sending resource updates to a peer node.
+type ReplicatorProtocol struct {
+	*baseProto
+	pushLogFunc           pushLogFunc
+	replicatorFailureFunc replicatorFailureFunc
+}
+
+// NewReplicatorProtocol returns and a new [ReplicatorProtocol] struct and registers the protocol
+// on the stream handler.
+func NewReplicatorProtocol(
+	h host.Host,
+	pushLogFunc pushLogFunc,
+	replicatorFailureFunc replicatorFailureFunc,
+) *ReplicatorProtocol {
+	proto := &ReplicatorProtocol{
+		baseProto:             newBaseProto(h),
+		pushLogFunc:           pushLogFunc,
+		replicatorFailureFunc: replicatorFailureFunc,
+	}
+	h.SetStreamHandler(replicatorProtocolRequest, proto.onRequest)
+	h.SetStreamHandler(replicatorProtocolResponse, proto.onResponse)
+	return proto
+}
+
+// PushToReplicator sends the pushlog request to the provided peer node.
+//
+// Callers should set an appropriate context timeout.
+func (proto *ReplicatorProtocol) PushToReplicator(
+	ctx context.Context,
+	evt event.Update,
+	pid peer.ID,
+) (reply *PushLogReply, err error) {
```

Comment:
thought: `reply` is never used by us, although it is exposed by the embedding of `*ReplicatorProtocol` in `server`.  Might be worth removing the return param.